src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -864,7 +898,437 @@ - (void)test005_DoBDXTransferAllowUpdateRequest
+    sOTAProviderDelegate.queryImageHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+        MTROTASoftwareUpdateProviderClusterQueryImageParams * params, QueryImageCompletion completion) {
+        XCTAssertEqualObjects(nodeID, @(kDeviceId1));
+        XCTAssertEqual(controller, sController);
+
+        sOTAProviderDelegate.queryImageHandler = nil;
+        [sOTAProviderDelegate respondAvailableWithDelay:@(0)
+                                                    uri:imagePath1
+                                            updateToken:updateToken
+                                        softwareVersion:kUpdatedSoftwareVersion2
+                                  softwareVersionString:kUpdatedSoftwareVersionString2
+                                             completion:completion];
+        [queryExpectation1 fulfill];
+    };
+
+    // Set up the BDX transfer begin, block query and transfer end handlers for device1 so it can go ahead with the BDX transfer
+    sOTAProviderDelegate.transferBeginHandler = ^(NSNumber * nodeID, MTRDeviceController * controller, NSString * fileDesignator,
+        NSNumber * offset, MTRStatusCompletion completion) {
+        XCTAssertEqualObjects(nodeID, @(kDeviceId1));
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(fileDesignator, imagePath1);
+        XCTAssertEqualObjects(offset, @(0));
+
+        readHandle = [NSFileHandle fileHandleForReadingAtPath:fileDesignator];
+        XCTAssertNotNil(readHandle);
+
+        NSError * endSeekError;
+        XCTAssertTrue([readHandle seekToEndReturningOffset:&imageSize error:&endSeekError]);
+        XCTAssertNil(endSeekError);
+
+        // Set up the query handler for device2 to return image available at imagePath2 but a busy delay of 30 secs
+        sOTAProviderDelegate.queryImageHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+            MTROTASoftwareUpdateProviderClusterQueryImageParams * params, QueryImageCompletion completion) {
+            XCTAssertEqualObjects(nodeID, @(kDeviceId2));
+            XCTAssertEqual(controller, sController);
+
+            sOTAProviderDelegate.queryImageHandler = nil;
+            [sOTAProviderDelegate respondAvailableWithDelay:@(busyDelay)
+                                                        uri:imagePath2
+                                                updateToken:updateToken1
+                                            softwareVersion:kUpdatedSoftwareVersion1
+                                      softwareVersionString:kUpdatedSoftwareVersionString1
+                                                 completion:completion];
+            [queryExpectation2 fulfill];
+        };
+        sOTAProviderDelegate.transferBeginHandler = nil;
+        [sOTAProviderDelegate respondSuccess:completion];
+        [bdxBeginExpectation1 fulfill];
+
+        // Announce ourselves to device2
+        announceResponseExpectation2 = [self announceProviderToDevice:device2];
```

Comment:
OK, so at this point the BDX transfer is racing with device2 doing its query, right?

_And_ the BDX transfer is racing with device 2 doing its followup query.  What happens if the BDX transfer takes longer than 30 seconds?

1. The `queryImageHandler` we install above should not set `sOTAProviderDelegate.queryImageHandler`.  It should just stick around and keep retrying busy.  _However_, it should fulfill `queryExpectation2` only once.  That will solve the "BDX takes longer than 30s" problem.
2. The `respondSuccess` bit should move into the query handler, similar to how `test003_ReceiveSecondQueryWhileHandlingBDX` does things, so that we ensure that the BDX transfer is still ongoing when the second query comes in.  That should happen only once, obviously, when `queryExpectation2` gets fulfilled.

Reply:
done.